File: analise/analise.R
# =============================================================
# Análise de Dados — Avaliação de Usabilidade Spotify Mobile
# Pipeline conforme Território n. 04 (Sergio Freitas) e exemplo.R
# Equipe 01 — IHC 2026/1
# =============================================================

# ---- 0. Setup ----
pkgs <- c("psychometric", "arules")
for (p in pkgs) {
  if (!requireNamespace(p, quietly = TRUE)) {
    install.packages(p, repos = "https://cloud.r-project.org", quiet = TRUE)
  }
}

# Caminho fixo p/ rodar via Rscript
setwd("/Users/vitormarconi/Documents/GitHub/ihc-spotify/docs/analise")

# ---- 1. Carga dos dados (separador ; como no exemplo.R do professor) ----
TodosDados.df <- read.csv2("dados.csv", stringsAsFactors = FALSE)
cat("=== AMOSTRA ===\n")
cat("Respondentes válidos:", nrow(TodosDados.df), "\n")
cat("Variáveis no arquivo:", ncol(TodosDados.df), "\n\n")

# ---- 2. Definição das colunas Likert e dos blocos por heurística ----
colunas.var <- c(
  # Bloco A — H3 Controle e Liberdade
  "A1_VoltarTela","A2_ControleRepr","A3_DesfazerPlay","A4_CancelarAcao","A5_RespeitaConfig",
  # Bloco B — H4 Consistência e Padrões
  "B1_SemelhApps","B2_IconesEsperados","B3_LayoutConsist","B4_AcharAposUpd","B5_PrevisivBotoes",
  # Bloco C — H5 Prevenção de Erros
  "C1_SemAcidente","C2_ControlesPos","C3_SugestoesBusca","C4_AvisoPerda","C5_TocaRapido",
  # Bloco D — H7 Flexibilidade e Eficiência
  "D1_AcoesRapidas","D2_Recomendacoes","D3_GerirPlaylists","D4_AdaptaUso","D5_TelasLimpas",
  # Bloco E — H8 Estética e Design Minimalista
  "E1_SemDistracao","E2_InfoVisivel","E3_Estetica","E4_HomeRelevante"
)

blocos <- list(
  H3_Controle        = colunas.var[1:5],
  H4_Consistencia    = colunas.var[6:10],
  H5_PrevencaoErros  = colunas.var[11:15],
  H7_Flexibilidade   = colunas.var[16:20],
  H8_Estetica        = colunas.var[21:24]
)

# Garantir tipo numérico (Likert 1–5)
for (c in colunas.var) TodosDados.df[[c]] <- as.numeric(TodosDados.df[[c]])

# ---- 3. Estatística descritiva por item ----
cat("=== DESCRITIVA POR ITEM ===\n")
desc.item <- data.frame(
  Item = colunas.var,
  N    = sapply(colunas.var, function(c) sum(!is.na(TodosDados.df[[c]]))),
  Media= round(sapply(colunas.var, function(c) mean(TodosDados.df[[c]], na.rm = TRUE)), 2),
  DP   = round(sapply(colunas.var, function(c) sd(TodosDados.df[[c]],   na.rm = TRUE)), 2),
  Mediana = sapply(colunas.var, function(c) median(TodosDados.df[[c]], na.rm = TRUE)),
  Pct_Baixo_1_2 = round(sapply(colunas.var, function(c) mean(TodosDados.df[[c]] <= 2, na.rm = TRUE)) * 100, 1),
  Pct_Alto_4_5  = round(sapply(colunas.var, function(c) mean(TodosDados.df[[c]] >= 4, na.rm = TRUE)) * 100, 1)
)
print(desc.item, row.names = FALSE)
write.csv2(desc.item, "saida_descritiva_itens.csv", row.names = FALSE)

# ---- 3b. Descritiva por bloco (score = média dos itens do bloco por respondente) ----
cat("\n=== DESCRITIVA POR BLOCO ===\n")
bloco.scores <- as.data.frame(lapply(blocos, function(cols) {
  rowMeans(TodosDados.df[, cols], na.rm = TRUE)
}))
desc.bloco <- data.frame(
  Bloco   = names(bloco.scores),
  Media   = round(sapply(bloco.scores, mean,   na.rm = TRUE), 2),
  DP      = round(sapply(bloco.scores, sd,     na.rm = TRUE), 2),
  Mediana = round(sapply(bloco.scores, median, na.rm = TRUE), 2),
  Min     = round(sapply(bloco.scores, min,    na.rm = TRUE), 2),
  Max     = round(sapply(bloco.scores, max,    na.rm = TRUE), 2),
  Pct_Baixo = round(sapply(bloco.scores, function(x) mean(x <= 2.5, na.rm = TRUE)) * 100, 1),
  Pct_Alto  = round(sapply(bloco.scores, function(x) mean(x >= 3.5, na.rm = TRUE)) * 100, 1)
)
print(desc.bloco, row.names = FALSE)
write.csv2(desc.bloco, "saida_descritiva_blocos.csv", row.names = FALSE)

# ---- 4. Alpha de Cronbach (global e por bloco) ----
cat("\n=== ALPHA DE CRONBACH ===\n")
library(psychometric)

interp_alpha <- function(a) {
  if (is.na(a))     return("NA")
  if (a >= 0.9)     return("EXCELENTE")
  if (a >= 0.8)     return("BOM")
  if (a >= 0.7)     return("ACEITAVEL")
  if (a >= 0.6)     return("QUESTIONAVEL")
  if (a >= 0.5)     return("POBRE")
  return("INACEITAVEL")
}

a.global <- alpha(TodosDados.df[colunas.var])
cat(sprintf("Global (24 itens):       alpha = %.3f  -> %s\n",
            a.global, interp_alpha(a.global)))

alpha.blocos <- sapply(blocos, function(cols) alpha(TodosDados.df[cols]))
for (b in names(alpha.blocos)) {
  cat(sprintf("%-22s alpha = %.3f  -> %s\n", b, alpha.blocos[b], interp_alpha(alpha.blocos[b])))
}

# ---- 5. Correlação de Pearson (entre blocos) ----
cat("\n=== CORRELACAO DE PEARSON (entre BLOCOS) ===\n")
V_Cor.bloco <- cor(bloco.scores, use = "pairwise.complete.obs", method = "pearson")
print(round(V_Cor.bloco, 3))
cat("\nsymnum (limiares 0.3/0.6/0.8/0.9/0.95):\n")
print(symnum(V_Cor.bloco))
write.csv2(round(V_Cor.bloco, 3), "saida_corr_blocos.csv")

interp_r <- function(r) {
  a <- abs(r)
  if (a > 0.9) return("Muito forte")
  if (a > 0.7) return("Forte")
  if (a > 0.5) return("Moderada")
  if (a > 0.3) return("Fraca")
  return("Desprezivel")
}
cat("\nPares de blocos ordenados por |r|:\n")
pares <- data.frame()
nm <- colnames(V_Cor.bloco)
for (i in 1:(length(nm)-1)) for (j in (i+1):length(nm)) {
  r <- V_Cor.bloco[i,j]
  pares <- rbind(pares, data.frame(A=nm[i], B=nm[j], r=round(r,3), interpretacao=interp_r(r)))
}
pares <- pares[order(-abs(pares$r)), ]
print(pares, row.names = FALSE)

# ---- 5b. Correlação Pearson item-a-item ----
cat("\n=== CORRELACOES FORTES ENTRE ITENS (|r| >= 0.5) ===\n")
V_Cor.item <- cor(TodosDados.df[colunas.var], use = "pairwise.complete.obs", method = "pearson")
fortes <- data.frame()
for (i in 1:(length(colunas.var)-1)) for (j in (i+1):length(colunas.var)) {
  r <- V_Cor.item[i,j]
  if (!is.na(r) && abs(r) >= 0.5) {
    fortes <- rbind(fortes,
      data.frame(Item_A = colunas.var[i], Item_B = colunas.var[j], r = round(r,3)))
  }
}
fortes <- fortes[order(-abs(fortes$r)), ]
print(head(fortes, 20), row.names = FALSE)
write.csv2(V_Cor.item, "saida_corr_itens.csv")
write.csv2(fortes, "saida_corr_fortes.csv", row.names = FALSE)

# ---- 6. Apriori — regras de associação ----
cat("\n=== APRIORI — REGRAS DE ASSOCIACAO ===\n")
library(arules)

# Discretiza Likert: 1-2=Baixo / 3=Neutro / 4-5=Alto (preserva NA)
discretiza <- function(v) {
  out <- rep(NA_character_, length(v))
  out[v <= 2 & !is.na(v)] <- "Baixo"
  out[v == 3 & !is.na(v)] <- "Neutro"
  out[v >= 4 & !is.na(v)] <- "Alto"
  factor(out, levels = c("Baixo","Neutro","Alto"))
}

# Versão 1: regras sobre os 5 BLOCOS (mais interpretável)
bloco.disc <- as.data.frame(lapply(bloco.scores, function(x) discretiza(round(x))))
g.bloco <- as(bloco.disc, "transactions")
r.bloco <- apriori(g.bloco,
                   parameter = list(supp = 0.10, conf = 0.60, minlen = 2, maxlen = 4),
                   control = list(verbose = FALSE))
cat(sprintf("Regras (blocos): %d  | supp >= 0.10, conf >= 0.60\n", length(r.bloco)))
r.bloco.sorted <- sort(r.bloco, by = "lift", decreasing = TRUE)
cat("\nTop 15 regras entre BLOCOS por LIFT:\n")
inspect(head(r.bloco.sorted, 15))
write(r.bloco.sorted, file = "saida_apriori_blocos.csv", sep = ";", quote = TRUE, row.names = FALSE)

# Versão 2: regras sobre ITENS (granularidade fina) — exigência maior de suporte
itens.disc <- as.data.frame(lapply(TodosDados.df[colunas.var], discretiza))
g.item <- as(itens.disc, "transactions")
r.item <- apriori(g.item,
                  parameter = list(supp = 0.20, conf = 0.80, minlen = 2, maxlen = 3),
                  control = list(verbose = FALSE))
cat(sprintf("\nRegras (itens): %d  | supp >= 0.20, conf >= 0.80\n", length(r.item)))
r.item.sorted <- sort(r.item, by = "lift", decreasing = TRUE)
cat("\nTop 20 regras entre ITENS por LIFT:\n")
inspect(head(r.item.sorted, 20))
write(r.item.sorted, file = "saida_apriori_itens.csv", sep = ";", quote = TRUE, row.names = FALSE)

# ---- 7. Demografia ----
cat("\n=== DEMOGRAFIA DA AMOSTRA VALIDA ===\n")
for (col in c("Idade","Escolaridade","Frequencia","Horas","Conta")) {
  cat("\n--", col, "--\n")
  tb <- sort(table(TodosDados.df[[col]], useNA = "ifany"), decreasing = TRUE)
  pct <- round(prop.table(tb) * 100, 1)
  print(data.frame(N = as.integer(tb), Pct = as.numeric(pct)))
}

cat("\n[ok] Análise concluída. Arquivos saida_*.csv gravados em ", getwd(), "\n", sep="")

File: analise/analise.py
"""
Análise de Dados — Avaliação de Usabilidade Spotify Mobile
Equivalente ao exemplo.R (Território n. 04 — Sergio Freitas / IHC)
Pipeline: Descritiva -> Alpha de Cronbach -> Correlação Pearson -> Apriori
"""
import pandas as pd
import numpy as np
from itertools import combinations
from collections import Counter
import json, os

BASE = '/Users/vitormarconi/Documents/GitHub/ihc-spotify/docs/analise'
os.chdir(BASE)

# ============ 1. CARGA ============
df = pd.read_csv('dados.csv', sep=';')
print(f"=== AMOSTRA ===")
print(f"Respondentes válidos: {len(df)}")
print(f"Variáveis no arquivo: {df.shape[1]}\n")

colunas_var = [
    'A1_VoltarTela','A2_ControleRepr','A3_DesfazerPlay','A4_CancelarAcao','A5_RespeitaConfig',
    'B1_SemelhApps','B2_IconesEsperados','B3_LayoutConsist','B4_AcharAposUpd','B5_PrevisivBotoes',
    'C1_SemAcidente','C2_ControlesPos','C3_SugestoesBusca','C4_AvisoPerda','C5_TocaRapido',
    'D1_AcoesRapidas','D2_Recomendacoes','D3_GerirPlaylists','D4_AdaptaUso','D5_TelasLimpas',
    'E1_SemDistracao','E2_InfoVisivel','E3_Estetica','E4_HomeRelevante',
]
blocos = {
    'H3_Controle':       colunas_var[0:5],
    'H4_Consistencia':   colunas_var[5:10],
    'H5_PrevencaoErros': colunas_var[10:15],
    'H7_Flexibilidade':  colunas_var[15:20],
    'H8_Estetica':       colunas_var[20:24],
}
for c in colunas_var:
    df[c] = pd.to_numeric(df[c], errors='coerce')

# ============ 2. DESCRITIVA POR ITEM ============
print("=== DESCRITIVA POR ITEM ===")
desc_rows = []
for bloco, cols in blocos.items():
    for c in cols:
        s = df[c].dropna()
        desc_rows.append({
            'Bloco': bloco, 'Item': c, 'N': len(s),
            'Media': round(s.mean(),2), 'DP': round(s.std(),2),
            'Mediana': s.median(),
            'Pct_Baixo_1_2': round((s<=2).mean()*100,1),
            'Pct_Neutro_3':  round((s==3).mean()*100,1),
            'Pct_Alto_4_5':  round((s>=4).mean()*100,1),
        })
desc = pd.DataFrame(desc_rows)
print(desc.to_string(index=False))
desc.to_csv('saida_descritiva_itens.csv', sep=';', index=False)

# ============ 2b. DESCRITIVA POR BLOCO ============
print("\n=== DESCRITIVA POR BLOCO (score = média dos itens) ===")
bloco_scores = pd.DataFrame({b: df[c].mean(axis=1, skipna=True) for b, c in blocos.items()})
desc_b = pd.DataFrame({
    'Bloco': list(blocos.keys()),
    'Media':   [round(bloco_scores[b].mean(),2)   for b in blocos],
    'DP':      [round(bloco_scores[b].std(),2)    for b in blocos],
    'Mediana': [round(bloco_scores[b].median(),2) for b in blocos],
    'Pct_Baixo_<=2.5': [round((bloco_scores[b]<=2.5).mean()*100,1) for b in blocos],
    'Pct_Alto_>=3.5':  [round((bloco_scores[b]>=3.5).mean()*100,1) for b in blocos],
})
print(desc_b.to_string(index=False))
desc_b.to_csv('saida_descritiva_blocos.csv', sep=';', index=False)

# ============ 3. ALPHA DE CRONBACH ============
def cronbach_alpha(d: pd.DataFrame):
    d = d.dropna(how='any')   # listwise deletion (como `psychometric::alpha`)
    k = d.shape[1]
    if d.shape[0] < 2 or k < 2: return float('nan'), 0
    var_items = d.var(axis=0, ddof=1).sum()
    var_total = d.sum(axis=1).var(ddof=1)
    if var_total == 0: return float('nan'), d.shape[0]
    return (k/(k-1)) * (1 - var_items/var_total), d.shape[0]

def interp_alpha(a):
    if pd.isna(a): return 'NA'
    if a >= 0.9: return 'EXCELENTE'
    if a >= 0.8: return 'BOM'
    if a >= 0.7: return 'ACEITAVEL'
    if a >= 0.6: return 'QUESTIONAVEL'
    if a >= 0.5: return 'POBRE'
    return 'INACEITAVEL'

print("\n=== ALPHA DE CRONBACH ===")
a_glob, n_glob = cronbach_alpha(df[colunas_var])
print(f"Global (24 itens):       alpha = {a_glob:.3f}  (n={n_glob} listwise)  -> {interp_alpha(a_glob)}")
alpha_blocos = {}
for b, cols in blocos.items():
    a, n = cronbach_alpha(df[cols])
    alpha_blocos[b] = a
    print(f"{b:<22} alpha = {a:.3f}  (n={n})         -> {interp_alpha(a)}")

# ============ 4. CORRELAÇÃO DE PEARSON ============
print("\n=== CORRELACAO DE PEARSON entre BLOCOS ===")
corr_b = bloco_scores.corr(method='pearson')
print(corr_b.round(3).to_string())
corr_b.round(3).to_csv('saida_corr_blocos.csv', sep=';')

def symnum(r):
    a = abs(r)
    if a >= 0.95: return 'B'
    if a >= 0.9:  return '*'
    if a >= 0.8:  return '+'
    if a >= 0.6:  return ','
    if a >= 0.3:  return '.'
    return ' '

print("\nsymnum (limiares: . 0.3 / , 0.6 / + 0.8 / * 0.9 / B 0.95):")
sym = corr_b.applymap(symnum)
print(sym.to_string())

def interp_r(r):
    a = abs(r)
    if a > 0.9: return 'Muito forte'
    if a > 0.7: return 'Forte'
    if a > 0.5: return 'Moderada'
    if a > 0.3: return 'Fraca'
    return 'Desprezivel'

print("\nPares de blocos ordenados por |r|:")
for a,b in sorted(combinations(corr_b.columns,2), key=lambda p: -abs(corr_b.loc[p[0],p[1]])):
    r = corr_b.loc[a,b]
    print(f"  {a:<22} ~ {b:<22} r={r:+.3f}  ({interp_r(r)})")

print("\n=== TOP CORRELACOES ITEM-A-ITEM (|r| >= 0.5) ===")
item_corr = df[colunas_var].corr(method='pearson')
strong = []
for a,b in combinations(colunas_var, 2):
    r = item_corr.loc[a,b]
    if abs(r) >= 0.5: strong.append((a, b, r))
strong.sort(key=lambda x: -abs(x[2]))
for a,b,r in strong[:20]:
    print(f"  {a:<22} ~ {b:<22} r={r:+.3f}")
pd.DataFrame(strong, columns=['Item_A','Item_B','r']).to_csv('saida_corr_fortes.csv', sep=';', index=False)

# ============ 5. APRIORI ============
def discretize(v):
    if pd.isna(v): return None
    if v <= 2: return 'Baixo'
    if v == 3: return 'Neutro'
    return 'Alto'

def apriori(items_per_tx, min_sup, min_conf, max_k=3):
    N = len(items_per_tx)
    def sup_count(itemset):
        return sum(1 for t in items_per_tx if itemset.issubset(t))
    # k=1
    c1 = Counter()
    for t in items_per_tx:
        for x in t: c1[x] += 1
    freq = {frozenset([x]): c/N for x,c in c1.items() if c/N >= min_sup}
    all_freq = dict(freq)
    prev = freq
    for k in range(2, max_k+1):
        items = sorted({i for s in prev for i in s})
        new = {}
        for combo in combinations(items, k):
            fs = frozenset(combo)
            s = sup_count(fs)/N
            if s >= min_sup: new[fs] = s
        all_freq.update(new)
        prev = new
        if not new: break
    rules = []
    for itemset, sup in all_freq.items():
        if len(itemset) < 2: continue
        for r_size in range(1, len(itemset)):
            for rhs_c in combinations(itemset, r_size):
                rhs = frozenset(rhs_c); lhs = itemset - rhs
                if not lhs: continue
                sup_lhs = all_freq.get(lhs); sup_rhs = all_freq.get(rhs)
                if sup_lhs is None or sup_rhs is None: continue
                conf = sup / sup_lhs
                lift = conf / sup_rhs
                if conf >= min_conf:
                    rules.append((tuple(sorted(lhs)), tuple(sorted(rhs)), sup, conf, lift))
    return rules

print("\n=== APRIORI — REGRAS DE ASSOCIACAO (BLOCOS) ===")
tx_bloco = []
for _, row in bloco_scores.iterrows():
    items = set()
    for b in bloco_scores.columns:
        v = row[b]
        if pd.isna(v): continue
        d = discretize(round(v))
        if d: items.add(f"{b}={d}")
    if items: tx_bloco.append(items)

rules_b = apriori(tx_bloco, min_sup=0.10, min_conf=0.60, max_k=3)
print(f"Transações: {len(tx_bloco)} | min_sup=0.10, min_conf=0.60 | regras: {len(rules_b)}")
print("\nTop 20 regras (blocos) por LIFT:")
for l, r, s, c, lf in sorted(rules_b, key=lambda x: -x[4])[:20]:
    print(f"  {{{', '.join(l)}}} => {{{', '.join(r)}}}  sup={s:.2f} conf={c:.2f} lift={lf:.2f}")
pd.DataFrame([{'se':' & '.join(l),'entao':' & '.join(r),'sup':round(s,3),'conf':round(c,3),'lift':round(lf,3)}
              for l,r,s,c,lf in sorted(rules_b, key=lambda x:-x[4])]).to_csv('saida_apriori_blocos.csv', sep=';', index=False)

print("\n=== APRIORI — REGRAS DE ASSOCIACAO (ITENS) ===")
tx_item = []
for _, row in df[colunas_var].iterrows():
    items = set()
    for c in colunas_var:
        d = discretize(row[c])
        if d: items.add(f"{c}={d}")
    if items: tx_item.append(items)

rules_i = apriori(tx_item, min_sup=0.20, min_conf=0.80, max_k=3)
print(f"Transações: {len(tx_item)} | min_sup=0.20, min_conf=0.80 | regras: {len(rules_i)}")
print("\nTop 25 regras (itens) por LIFT:")
for l, r, s, c, lf in sorted(rules_i, key=lambda x: -x[4])[:25]:
    print(f"  {{{', '.join(l)}}} => {{{', '.join(r)}}}  sup={s:.2f} conf={c:.2f} lift={lf:.2f}")
pd.DataFrame([{'se':' & '.join(l),'entao':' & '.join(r),'sup':round(s,3),'conf':round(c,3),'lift':round(lf,3)}
              for l,r,s,c,lf in sorted(rules_i, key=lambda x:-x[4])]).to_csv('saida_apriori_itens.csv', sep=';', index=False)

# ============ 6. DEMOGRAFIA ============
print("\n=== PERFIL DEMOGRAFICO ===")
for col, name in [('Idade','Faixa etária'),('Escolaridade','Escolaridade'),
                  ('Frequencia','Frequência de uso'),('Horas','Horas por semana'),('Conta','Tipo de conta')]:
    print(f"\n-- {name} --")
    vc = df[col].value_counts(dropna=False)
    for k,v in vc.items():
        print(f"  {str(k)[:55]:<57} N={v:>3}  ({v/len(df)*100:5.1f}%)")

# ============ SALVAR JSON RESUMO ============
resumo = {
    'n_validos': int(len(df)),
    'alpha_global': round(a_glob,3),
    'alpha_blocos': {k: round(v,3) for k,v in alpha_blocos.items()},
    'desc_blocos':  desc_b.to_dict(orient='records'),
    'corr_blocos':  corr_b.round(3).to_dict(),
    'top_regras_blocos': [
        {'se':list(l),'entao':list(r),'sup':round(s,3),'conf':round(c,3),'lift':round(lf,3)}
        for l,r,s,c,lf in sorted(rules_b, key=lambda x:-x[4])[:15]],
    'top_regras_itens': [
        {'se':list(l),'entao':list(r),'sup':round(s,3),'conf':round(c,3),'lift':round(lf,3)}
        for l,r,s,c,lf in sorted(rules_i, key=lambda x:-x[4])[:15]],
}
with open('saida_resumo.json','w') as f:
    json.dump(resumo, f, ensure_ascii=False, indent=2, default=str)
print(f"\n[ok] Resultados salvos em {BASE}/saida_*.csv|json")
